src/Compilers/CSharp/Test/Semantic/Semantics/LambdaTests.cs

```diff
@@ -4329,6 +4329,584 @@ static void Main()
+                // (8,29): error CS1073: Unexpected token 'ref'
+                //         Delegate d1 = async ref Task (s) => { _ = s.Length; await Task.Yield(); };
+                Diagnostic(ErrorCode.ERR_UnexpectedToken, "ref").WithArguments("ref").WithLocation(8, 29),
+                // (9,22): error CS1073: Unexpected token 'ref'
+                //         D d2 = async ref Task (s) => { _ = s.Length; await Task.Yield(); };
+                Diagnostic(ErrorCode.ERR_UnexpectedToken, "ref").WithArguments("ref").WithLocation(9, 22));
+        }
+
+        [Fact]
+        public void LambdaReturnType_14()
+        {
+            var source =
+@"using System;
+using System.Threading.Tasks;
+delegate ref Task D(string s);
+class Program
+{
+    static void Main()
+    {
+        Delegate d1 = async ref Task (string s) => { _ = s.Length; await Task.Yield(); };
+        D d2 = async ref Task (string s) => { _ = s.Length; await Task.Yield(); };
+    }
+}";
+            var comp = CreateCompilation(source, parseOptions: TestOptions.RegularPreview);
+            comp.VerifyDiagnostics(
+                // (8,23): error CS8917: The delegate type could not be inferred.
+                //         Delegate d1 = async ref Task (string s) => { _ = s.Length; await Task.Yield(); };
+                Diagnostic(ErrorCode.ERR_CannotInferDelegateType, "async ref Task (string s) => { _ = s.Length; await Task.Yield(); }").WithLocation(8, 23),
+                // (8,29): error CS1073: Unexpected token 'ref'
+                //         Delegate d1 = async ref Task (string s) => { _ = s.Length; await Task.Yield(); };
+                Diagnostic(ErrorCode.ERR_UnexpectedToken, "ref").WithArguments("ref").WithLocation(8, 29),
+                // (9,22): error CS1073: Unexpected token 'ref'
+                //         D d2 = async ref Task (string s) => { _ = s.Length; await Task.Yield(); };
+                Diagnostic(ErrorCode.ERR_UnexpectedToken, "ref").WithArguments("ref").WithLocation(9, 22));
+        }
+
+        [Fact]
+        public void LambdaReturnType_UseSiteErrors()
+        {
+            var sourceA =
+@".class public sealed A extends [mscorlib]System.Object
+{
+  .custom instance void [mscorlib]System.Runtime.CompilerServices.RequiredAttributeAttribute::.ctor(class [mscorlib]System.Type) = ( 01 00 FF 00 00 ) 
+  .method public hidebysig specialname rtspecialname instance void .ctor() cil managed { ret }
+}";
+            var refA = CompileIL(sourceA);
+
+            var sourceB =
+@"using System;
+class B
+{
+    static void F<T>(Func<T> f) { }
+    static void Main()
+    {
+        F(A () => default);
+    }
+}";
+            var comp = CreateCompilation(sourceB, references: new[] { refA }, parseOptions: TestOptions.RegularPreview);
+            comp.VerifyDiagnostics(
+                // (7,11): error CS0648: 'A' is a type not supported by the language
```

Comment:
This isn't the diagnostic I expected for a use-site issue, this looks more like bad PE. Should the test name be adjusted?
If so, could you add a use-site error scenario?

Reply:
This is a use-site diagnostic reported in `PENamedTypeSymbol.GetUseSiteDiagnosticImpl()`.